jwt/jwt.go

```diff
@@ -0,0 +1,128 @@
+package jwt
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+type header struct {
+	// The algorithm used for signature.
+	Algorithm string `json:"alg"`
+	// Represents the token type.
+	Typ string `json:"typ"`
+	// The hint which key is being used.  ID Tokens SHOULD NOT use the JWS or
+	// JWE x5u, x5c, jku, or jwk Header Parameter fields. Instead, references to
+	// keys used are communicated in advance using Discovery and Registration
+	// parameters, per Section 10.
+	KeyID string `json:"kid"`
+}
+
+type ClaimSet struct {
+	Issuer    string   `json:"iss,omitempty"`
+	Subject   string   `json:"sub,omitempty"`
+	Audience  []string `json:"aud,omitempty"`
+	Expiry    int64    `json:"exp,omitempty"`
+	NotBefore int64    `json:"nbf,omitempty"`
+	IssuedAt  int64    `json:"iat,omitempty"`
+	JwtId     string   `json:"jti,omitempty"`
+}
+
+func EncodeAndSign(claims interface{}, keyID string, privateKey *ecdsa.PrivateKey) (string, error) {
+	header := header{
+		Algorithm: "ES256",
+		Typ:       "JWT",
+		KeyID:     keyID,
+	}
+	headerJSON, err := json.Marshal(header)
+	if err != nil {
+		return "", err
+	}
+	headerB64 := base64.RawURLEncoding.EncodeToString(headerJSON)
+	claimsJSON, err := json.Marshal(claims)
+	if err != nil {
+		return "", err
+	}
+	claimsB64 := base64.RawURLEncoding.EncodeToString(claimsJSON)
+
+	toSign := strings.Join([]string{headerB64, claimsB64}, ".")
+	hash := sha256.Sum256([]byte(toSign))
+	r, s, err := ecdsa.Sign(rand.Reader, privateKey, hash[:])
+	if err != nil {
+		return "", err
+	}
+	curveBits := privateKey.Curve.Params().BitSize
+
+	keyBytes := curveBits / 8
+	if curveBits%8 > 0 {
+		keyBytes += 1
```

Comment:
`keyBytes++` is recommended over `+= 1` unless at some point that constant might change.